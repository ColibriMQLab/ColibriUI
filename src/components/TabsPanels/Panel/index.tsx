import React from "react";
import classNames from "classnames";
import type { PanelProps } from "./index.props";
import type { FC } from "react";

export const TabsPanel: FC<PanelProps> = ({
  content,
  className,
  id,
  ...props
}) => (
  <div
    {...props}
    id={id}
    role="tabpanel"
    className={classNames(null, className)}
  >
    {content}
  </div>
);
