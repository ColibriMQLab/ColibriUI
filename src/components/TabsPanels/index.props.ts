import type { RefObject } from "react";
import type { PanelProps } from "./Panel/index.props";

export type TabsPanelsProps = {
  activePanel?: string;
  panels: PanelProps[];
  innerRef?: RefObject<HTMLDivElement>;
  className?: string;
};
