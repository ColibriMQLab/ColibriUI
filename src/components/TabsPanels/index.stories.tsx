import React from 'react'
import TabsPanels from ".";
import type { Meta, StoryObj } from "@storybook/react-webpack5";

const meta: Meta<typeof TabsPanels> = {
  title: "UI/TabsPanels",
  parameters: {
    layout: "centered",
  },
  argTypes: {
  },
  component: TabsPanels,
} satisfies Meta<typeof TabsPanels>;

export default meta;

type Story = StoryObj<typeof TabsPanels>;

export const Default = (args: Story) => {
  return(
  <TabsPanels
    {...args}
    activePanel="2"
    panels={[
      { id: '1', content: 'Pane 1 content' },
      { id: '2', content: 'Pane 2 content' },
    ]}
  />
)};