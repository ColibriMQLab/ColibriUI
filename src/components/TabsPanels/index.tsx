import React from "react";

import classNames from "classnames";
import { TabsPanel } from "./Panel";
import type { TabsPanelsProps } from "./index.props";
import type { FC } from "react";
import type { PanelProps } from "./Panel/index.props";

const TabsPanels: FC<TabsPanelsProps> = ({
  activePanel,
  className,
  innerRef,
  panels,
  ...props
}) => {
  const currentPanel = panels.reduce<PanelProps | null>((acc, panel) => {
    if (acc !== null) {
      return acc;
    }
    return panel.id === activePanel ? panel : null;
  }, null);

  if (!currentPanel) {
    return null;
  }

  const { id, content, ...panelProps } = currentPanel;

  return (
    <div
      {...props}
      ref={innerRef}
      role="menutabpanels"
      className={classNames(null, [className])}
    >
      <TabsPanel
        aria-labelledby={`tab-${activePanel}`}
        id={id}
        content={content}
        {...panelProps}
      />
    </div>
  );
};

export default TabsPanels;
