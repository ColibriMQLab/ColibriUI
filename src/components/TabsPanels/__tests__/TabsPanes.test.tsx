import React, { createRef } from 'react';
import { render, screen } from '@testing-library/react';
import TabsPanels from '..';

const panel1 = { id: 'tab1', content: 'Pane 1 Поиск' };
const panel2 = { id: 'tab2', content: 'Pane 2 Картинки' };
const panel3 = { id: 'tab3', content: <div>Pane 3 Маркет</div> };

const panels = [panel1, panel2, panel3];

describe('TabsPanels', () => {
  test('should return the complete component template (snapshot)', () => {
    expect(render(<TabsPanels activePanel={panel1.id} panels={panels} className="mix" />)).toMatchSnapshot();
  });
  
  test('should set a reference to the root DOM element', async () => {
    const innerRef = createRef<HTMLDivElement>();
    render(<TabsPanels activePanel={panel1.id} panels={panels} innerRef={innerRef} />);
  
    expect(innerRef.current).not.toBe(null);
  });
  
  test('should display only the active pane content (snapshot)', () => {
    expect(render(<TabsPanels activePanel={panel1.id} panels={panels} />)).toMatchSnapshot();
  });
  
  test('should display new JSX content when activePane changes (snapshot)', () => {
    const { container, rerender } = render(<TabsPanels activePanel={panel1.id} panels={panels} />);
    rerender(<TabsPanels activePanel={panel3.id} panels={panels} />);
    expect(container).toMatchSnapshot();
  });
  
  test('should render an empty panel when activePane is not set', () => {
    render(<TabsPanels activePanel={undefined} panels={panels} />);
    expect(screen.queryByRole('menutabpanels')).toBeNull();
  
  });
});
