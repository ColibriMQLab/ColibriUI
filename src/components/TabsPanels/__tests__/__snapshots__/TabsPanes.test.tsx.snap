// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabsPanels should display new JSX content when activePane changes (snapshot) 1`] = `
<div>
  <div
    class=""
    role="menutabpanels"
  >
    <div
      aria-labelledby="tab-tab3"
      class=""
      id="tab3"
      role="tabpanel"
    >
      <div>
        Pane 3 Маркет
      </div>
    </div>
  </div>
</div>
`;

exports[`TabsPanels should display only the active pane content (snapshot) 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class=""
        role="menutabpanels"
      >
        <div
          aria-labelledby="tab-tab1"
          class=""
          id="tab1"
          role="tabpanel"
        >
          Pane 1 Поиск
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class=""
      role="menutabpanels"
    >
      <div
        aria-labelledby="tab-tab1"
        class=""
        id="tab1"
        role="tabpanel"
      >
        Pane 1 Поиск
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`TabsPanels should return the complete component template (snapshot) 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="mix"
        role="menutabpanels"
      >
        <div
          aria-labelledby="tab-tab1"
          class=""
          id="tab1"
          role="tabpanel"
        >
          Pane 1 Поиск
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="mix"
      role="menutabpanels"
    >
      <div
        aria-labelledby="tab-tab1"
        class=""
        id="tab1"
        role="tabpanel"
      >
        Pane 1 Поиск
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
